File: quant/kelly.py
#!/usr/bin/env python3
"""
Kelly criterion — single-bet, fractional, and multi-bet correlated (Markowitz-Kelly).

Supports both betting edges (decimal odds) and asset returns (mean/vol).

Usage (single bet):
    python3 kelly.py --mode single --p 0.55 --odds-decimal 2.0 --fraction 0.25

Usage (portfolio of bets, possibly correlated):
    python3 kelly.py --mode portfolio --edges-json edges.json --fraction 0.25
    # edges.json: [{"label":"NBA-BOS-ML","p":0.58,"odds":1.91},{...}]
"""
import argparse
try:
    from ._validation import number, series, confidence_level
except ImportError:
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    from _validation import number, series, confidence_level
import json
import math
import sys


def kelly_single(p, odds_decimal):
    """
    Optimal full-Kelly fraction for a binary bet at decimal odds.
    b = odds - 1 (net payout), q = 1 - p
    f = (bp - q) / b
    """
    number(p, "p", minimum=0, maximum=1)
    number(odds_decimal, "odds_decimal")
    if odds_decimal <= 1:
        raise ValueError("decimal odds must exceed 1")
    b = odds_decimal - 1
    q = 1 - p
    f = (b * p - q) / b if b > 0 else 0.0
    return max(0.0, f)


def edge_pct(p, odds_decimal):
    """Expected value as % of stake."""
    kelly_single(p, odds_decimal)
    return p * (odds_decimal - 1) - (1 - p)


def kelly_portfolio(edges, fraction=0.25, corr_matrix=None):
    """
    Naive multi-bet Kelly assuming independence (or low correlation).
    Scales each bet by fractional Kelly and caps total exposure.
    If corr_matrix provided, applies diversification shrinkage.
    """
    number(fraction, "fraction", minimum=0, maximum=1)
    labels = [e.get("label", f"bet_{i}") for i, e in enumerate(edges)]
    if len(set(labels)) != len(labels):
        raise ValueError("bet labels must be unique")
    if corr_matrix is not None:
        if not isinstance(corr_matrix, dict):
            raise ValueError("correlation matrix must map labels to rows")
        for label, row in corr_matrix.items():
            if label not in labels or not isinstance(row, dict):
                raise ValueError("correlation rows must use known bet labels")
            for other, correlation in row.items():
                if other not in labels:
                    raise ValueError("correlation columns must use known bet labels")
                number(correlation, "correlation", minimum=-1, maximum=1)
                if label == other and correlation != 1:
                    raise ValueError("correlation diagonal must equal 1")
                reverse = corr_matrix.get(other, {}).get(label)
                if reverse is not None and not math.isclose(correlation, reverse, abs_tol=1e-12):
                    raise ValueError("correlations must be symmetric when both entries are supplied")
    results = []
    total_raw = 0.0
    for e in edges:
        k = kelly_single(e["p"], e["odds"])
        ev = edge_pct(e["p"], e["odds"])
        stake = k * fraction
        results.append({
            "label": e.get("label", f"bet_{len(results)}"),
            "p": e["p"],
            "odds": e["odds"],
            "ev_pct": round(ev * 100, 3),
            "full_kelly_pct": round(k * 100, 3),
            "fractional_kelly_pct": round(stake * 100, 3),
        })
        total_raw += stake

    # Total exposure cap: sum of fractional Kellys shouldn't exceed 50% of bankroll
    if total_raw > 0.50:
        scale = 0.50 / total_raw
        for r in results:
            r["fractional_kelly_pct"] = round(r["fractional_kelly_pct"] * scale, 3)
            r["scaled_for_total_exposure"] = True

    # Correlation shrinkage: if provided, cut each stake proportional to its avg correlation to others
    if corr_matrix:
        for i, r in enumerate(results):
            row = corr_matrix.get(r["label"], {})
            if row:
                others = [v for k, v in row.items() if k != r["label"]]
                avg_corr = sum(others) / len(others) if others else 0.0
                shrink = max(0.5, 1 - max(0.0, avg_corr) * 0.5)  # haircut only; no leverage from negative correlations
                r["fractional_kelly_pct"] = round(r["fractional_kelly_pct"] * shrink, 3)
                r["correlation_shrink"] = round(shrink, 3)

    # Display rounding must not turn the stated 50% ceiling into an excess.
    excess_units = max(0, sum(round(r["fractional_kelly_pct"] * 1000) for r in results) - 50000)
    for r in sorted(results, key=lambda row: row["fractional_kelly_pct"], reverse=True):
        reduction = min(excess_units, round(r["fractional_kelly_pct"] * 1000))
        r["fractional_kelly_pct"] = round(r["fractional_kelly_pct"] - reduction / 1000, 3)
        excess_units -= reduction
    total_exposure = sum(r["fractional_kelly_pct"] for r in results)
    return {
        "fraction": fraction,
        "bets": results,
        "total_exposure_pct": round(total_exposure, 3),
        "n_bets": len(results),
        "diversification_benefit": "applied" if corr_matrix else "not_applied",
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--mode", choices=["single", "portfolio"], required=True)
    ap.add_argument("--p", type=float, help="win probability (single mode)")
    ap.add_argument("--odds-decimal", type=float, help="decimal odds (single mode)")
    ap.add_argument("--edges-json", help="path to edges JSON (portfolio mode)")
    ap.add_argument("--correlation-matrix", help="path to correlation matrix JSON (portfolio mode)")
    ap.add_argument("--fraction", type=float, default=0.25, help="Kelly fraction (default 0.25 = quarter)")
    args = ap.parse_args()

    number(args.fraction, "fraction", minimum=0, maximum=1)
    if args.mode == "single":
        if args.p is None or args.odds_decimal is None:
            print(json.dumps({"error": "need --p and --odds-decimal"}))
            sys.exit(1)
        k = kelly_single(args.p, args.odds_decimal)
        ev = edge_pct(args.p, args.odds_decimal)
        out = {
            "p": args.p,
            "odds": args.odds_decimal,
            "edge_pct": round(ev * 100, 3),
            "full_kelly_pct": round(k * 100, 3),
            "fractional_kelly_pct": round(k * args.fraction * 100, 3),
            "fraction": args.fraction,
        }
    else:
        if not args.edges_json:
            print(json.dumps({"error": "need --edges-json"}))
            sys.exit(1)
        with open(args.edges_json) as f:
            edges = json.load(f)
        corr = None
        if args.correlation_matrix:
            with open(args.correlation_matrix) as f:
                corr = json.load(f)
        out = kelly_portfolio(edges, args.fraction, corr)

    print(json.dumps(out, indent=2))


if __name__ == "__main__":
    main()

File: quant-jvm/src/main/kotlin/org/maxmoran/quant/Kelly.kt
package org.maxmoran.quant

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.math.max
import kotlin.system.exitProcess

/**
 * Kelly criterion — single-bet, fractional, and multi-bet correlated (Markowitz-Kelly).
 *
 * Parity contract: numerical agreement with [analyst-toolkit/quant/kelly.py] to within 1e-10
 * on the underlying math, and exact agreement on the percentage fields after Python's
 * `round(..., 3)` semantics (half-to-even, matched via BigDecimal HALF_EVEN here).
 *
 * Usage (single bet):
 *   ./gradlew run --args="kelly --mode single --p 0.55 --odds 2.0 --fraction 0.25"
 *
 * Usage (portfolio):
 *   ./gradlew run --args="kelly --mode portfolio --edges-json edges.json --fraction 0.25"
 */

/** A single betting opportunity. Mirrors the JSON schema accepted by `kelly.py`. */
data class Edge(val label: String, val p: Double, val odds: Double)

/** Internal per-bet result, mutable during exposure-cap + correlation-shrink adjustments. */
data class KellyRow(
    val label: String,
    val p: Double,
    val odds: Double,
    val evPct: Double,
    val fullKellyPct: Double,
    var fractionalKellyPct: Double,
    var scaledForTotalExposure: Boolean = false,
    var correlationShrink: Double? = null,
)

/** Optimal full-Kelly fraction for a binary bet at decimal odds. */
fun kellySingle(p: Double, oddsDecimal: Double): Double {
    require(p.isFinite() && p in 0.0..1.0) { "probability must be in [0, 1]" }
    require(oddsDecimal.isFinite() && oddsDecimal > 1.0) { "decimal odds must exceed 1" }
    val b = oddsDecimal - 1.0
    val q = 1.0 - p
    val f = if (b > 0.0) (b * p - q) / b else 0.0
    return max(0.0, f)
}

/** Expected value as % of stake (decimal, e.g. 0.052 = 5.2% edge). */
fun edgePct(p: Double, oddsDecimal: Double): Double {
    kellySingle(p, oddsDecimal)
    return p * (oddsDecimal - 1.0) - (1.0 - p)
}

/**
 * Naive multi-bet Kelly assuming low correlation by default. Optional correlation matrix
 * applies diversification shrinkage with a 50% floor. Total-exposure cap (50% of bankroll)
 * applies regardless. Sequencing matches `kelly.py` exactly: exposure cap first, then
 * correlation shrink.
 */
fun kellyPortfolio(
    edges: List<Edge>,
    fraction: Double = 0.25,
    corrMatrix: Map<String, Map<String, Double>>? = null,
): JsonObject {
    require(fraction.isFinite() && fraction in 0.0..1.0) { "fraction must be in [0, 1]" }
    val labels = edges.mapIndexed { i, edge -> edge.label.ifBlank { "bet_$i" } }
    require(labels.toSet().size == labels.size) { "bet labels must be unique" }
    corrMatrix?.forEach { (label, row) ->
        require(label in labels) { "unknown correlation row" }
        row.forEach { (other, correlation) ->
            require(other in labels && correlation.isFinite() && correlation in -1.0..1.0) { "invalid correlation" }
            require(label != other || correlation == 1.0) { "correlation diagonal must equal 1" }
            val reverse = corrMatrix[other]?.get(label)
            require(reverse == null || kotlin.math.abs(reverse - correlation) <= 1e-12) { "correlations must be symmetric" }
        }
    }
    val results = edges.mapIndexed { i, e ->
        val k = kellySingle(e.p, e.odds)
        val ev = edgePct(e.p, e.odds)
        KellyRow(
            label = e.label.ifBlank { "bet_$i" },
            p = e.p,
            odds = e.odds,
            evPct = round3(ev * 100.0),
            fullKellyPct = round3(k * 100.0),
            fractionalKellyPct = round3(k * fraction * 100.0),
        )
    }

    val totalRaw = edges.sumOf { kellySingle(it.p, it.odds) * fraction } * 100.0
    if (totalRaw > 50.0) {
        val scale = 50.0 / totalRaw
        results.forEach {
            it.fractionalKellyPct = round3(it.fractionalKellyPct * scale)
            it.scaledForTotalExposure = true
        }
    }

    if (corrMatrix != null) {
        results.forEach { r ->
            val row = corrMatrix[r.label] ?: return@forEach
            val others = row.filterKeys { it != r.label }
            if (row.isNotEmpty()) {
                val avgCorr = if (others.isEmpty()) 0.0 else others.values.sum() / others.size
                val shrink = max(0.5, 1.0 - max(0.0, avgCorr) * 0.5)
                r.fractionalKellyPct = round3(r.fractionalKellyPct * shrink)
                r.correlationShrink = round3(shrink)
            }
        }
    }

    var excessUnits = max(0L, results.sumOf { kotlin.math.round(it.fractionalKellyPct * 1000).toLong() } - 50000L)
    results.sortedByDescending { it.fractionalKellyPct }.forEach {
        val reduction = minOf(excessUnits, kotlin.math.round(it.fractionalKellyPct * 1000).toLong())
        it.fractionalKellyPct = round3(it.fractionalKellyPct - reduction / 1000.0)
        excessUnits -= reduction
    }
    val totalExposure = round3(results.sumOf { it.fractionalKellyPct })

    return buildJsonObject {
        put("fraction", JsonPrimitive(fraction))
        put("bets", buildJsonArray { results.forEach { add(rowToJson(it)) } })
        put("total_exposure_pct", JsonPrimitive(totalExposure))
        put("n_bets", JsonPrimitive(results.size))
        put("diversification_benefit",
            JsonPrimitive(if (corrMatrix != null) "applied" else "not_applied"))
    }
}

private fun rowToJson(r: KellyRow): JsonObject = buildJsonObject {
    put("label", JsonPrimitive(r.label))
    put("p", JsonPrimitive(r.p))
    put("odds", JsonPrimitive(r.odds))
    put("ev_pct", JsonPrimitive(r.evPct))
    put("full_kelly_pct", JsonPrimitive(r.fullKellyPct))
    put("fractional_kelly_pct", JsonPrimitive(r.fractionalKellyPct))
    if (r.scaledForTotalExposure) put("scaled_for_total_exposure", JsonPrimitive(true))
    r.correlationShrink?.let { put("correlation_shrink", JsonPrimitive(it)) }
}

/**
 * Match Python's `round(value, 3)` half-to-even (banker's rounding). Python uses HALF_EVEN
 * for `round()` on floats; we mirror via BigDecimal HALF_EVEN. NaN/Infinity are passed through.
 */
fun round3(value: Double): Double {
    if (value.isNaN() || value.isInfinite()) return value
    return java.math.BigDecimal(value).setScale(3, java.math.RoundingMode.HALF_EVEN).toDouble()
}

/** Module CLI runner — JSON in / JSON out, matches `kelly.py`'s contract. Invoked from Cli.kt. */
fun runKelly(args: Array<String>) {
    val parsed = parseArgs(args)
    val mode = parsed["mode"] ?: errorOut("need --mode single|portfolio")
    val json = Json { prettyPrint = true; encodeDefaults = false }

    val output: JsonElement = when (mode) {
        "single" -> {
            val p = parsed["p"]?.toDoubleOrNull() ?: errorOut("need --p")
            val odds = (parsed["odds-decimal"] ?: parsed["odds"])?.toDoubleOrNull()
                ?: errorOut("need --odds-decimal (or --odds)")
            val fraction = parsed["fraction"]?.toDoubleOrNull() ?: 0.25
            require(fraction.isFinite() && fraction in 0.0..1.0) { "fraction must be in [0, 1]" }
            val k = kellySingle(p, odds)
            val ev = edgePct(p, odds)
            buildJsonObject {
                put("p", JsonPrimitive(p))
                put("odds", JsonPrimitive(odds))
                put("edge_pct", JsonPrimitive(round3(ev * 100.0)))
                put("full_kelly_pct", JsonPrimitive(round3(k * 100.0)))
                put("fractional_kelly_pct", JsonPrimitive(round3(k * fraction * 100.0)))
                put("fraction", JsonPrimitive(fraction))
            }
        }
        "portfolio" -> {
            val edgesPath = parsed["edges-json"] ?: errorOut("need --edges-json")
            val fraction = parsed["fraction"]?.toDoubleOrNull() ?: 0.25
            val edges = parseEdgesJson(java.io.File(edgesPath).readText())
            val corr = parsed["correlation-matrix"]?.let {
                parseCorrJson(java.io.File(it).readText())
            }
            kellyPortfolio(edges, fraction, corr)
        }
        else -> errorOut("unknown mode: $mode")
    }
    println(json.encodeToString(JsonElement.serializer(), output))
}

private fun parseArgs(args: Array<String>): Map<String, String> {
    val m = mutableMapOf<String, String>()
    var i = 0
    while (i < args.size) {
        val a = args[i]
        if (a.startsWith("--")) {
            val key = a.removePrefix("--")
            val value = if (i + 1 < args.size && !args[i + 1].startsWith("--")) {
                i += 1; args[i]
            } else "true"
            m[key] = value
        }
        i += 1
    }
    return m
}

private fun parseEdgesJson(text: String): List<Edge> =
    Json.parseToJsonElement(text).jsonArray.map { el ->
        val o = el.jsonObject
        Edge(
            label = o["label"]?.jsonPrimitive?.content ?: "",
            p = o["p"]!!.jsonPrimitive.content.toDouble(),
            odds = o["odds"]!!.jsonPrimitive.content.toDouble(),
        )
    }

private fun parseCorrJson(text: String): Map<String, Map<String, Double>> =
    Json.parseToJsonElement(text).jsonObject.mapValues { (_, v) ->
        v.jsonObject.mapValues { (_, vv) -> vv.jsonPrimitive.content.toDouble() }
    }

private fun errorOut(msg: String): Nothing {
    println("""{"error": "$msg"}""")
    exitProcess(1)
}
